src/core/search/rax_tree.h

```diff
@@ -0,0 +1,149 @@
+#pragma once
+
+#include <absl/types/span.h>
+
+#include <cstdio>
+#include <memory_resource>
+#include <optional>
+#include <string_view>
+
+extern "C" {
+
+#include "redis/rax.h"
+}
+
+namespace dfly::search {
+
+// absl::flat_hash_map/std::unordered_map compatible tree map based on rax tree.
+// Allocates all objects on heap (with custom memory resource) as rax tree operates fully on
+// pointers.
+template <typename V> struct RaxTreeMap {
+  struct FindIterator;
+
+  // Simple seeking iterator
+  struct SeekIterator {
+    friend struct FindIterator;
+
+    SeekIterator() {
+    }
+
+    ~SeekIterator() {
+      if (it)
+        raxStop(&*it);
+    }
+
+    SeekIterator(rax* tree, const char* op, std::string_view key) {
+      it.emplace();
+      raxStart(&*it, tree);
+      raxSeek(&*it, op, to_key_ptr(key), key.size());
+      operator++();
+    }
+
+    explicit SeekIterator(rax* tree) : SeekIterator(tree, "^", std::string_view{nullptr, 0}) {
+    }
+
+    bool operator==(const SeekIterator& rhs) const {
+      return it.has_value() == rhs.it.has_value() && (!it.has_value() || it->key == rhs.it->key);
+    }
+
+    bool operator!=(const SeekIterator& rhs) const {
+      return !operator==(rhs);
+    }
+
+    SeekIterator& operator++() {
+      if (!raxNext(&*it)) {
+        raxStop(&*it);
+        it.reset();
+      }
+      return *this;
+    }
+
+    std::pair<std::string_view, V&> operator*() const {
+      return {std::string_view{reinterpret_cast<const char*>(it->key), it->key_len},
+              *reinterpret_cast<V*>(it->data)};
+    }
+
+   private:
+    std::optional<raxIterator> it;
```

Comment:
`it_`

Reply:
nit: raxIterator can easily support the "uninitialized" state without changing rax code, so optional here is redundant.